core-bundle/src/Migration/AbstractRecordedMigration.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Migration;
+
+use Contao\CoreBundle\Entity\Migration as MigrationEntity;;
+use Contao\CoreBundle\Repository\MigrationRepository;
+use Doctrine\DBAL\Connection;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Contracts\Service\ServiceSubscriberTrait;
+
+abstract class AbstractRecordedMigration extends AbstractMigration
+{
+    use ServiceSubscriberTrait;
+
+    protected function hasRun(): ?bool
+    {
+        if (!$this->connection()->getSchemaManager()->tablesExist(['tl_migration'])) {
+            return null;
+        }
+
+        return null !== $this->migrationRepository()->findOneByName($this->getName());
+    }
+
```

Comment:
You mean:

```php
abstract public function shouldRun(): bool
{
    return false !== $this->hasRun();
}
```

and in your own code:

```
public functon shouldRun(): bool
{
    // my conditions
    
    return parent::shouldRun();
}
```

?

Reply:
I wonder if we could simply replace the `shouldRun` method? And if you wanna extend it, you can call parent?

Reply:
> I wonder if we could simply replace the shouldRun method? And if you wanna extend it, you can call parent?

Yes that's what I would do. Just remove the `abstract` IMHO.

Reply:
As discussed in slack, in most cases you will need to implement your own `shouldRun` method anyway - unless your migration has nothing to do with the database at all. Otherwise you will always have to check whether your affected tables and fields are actually present (since the migration could be executed before any of them are there). Otherwise I am fine with provided a default implementation for the `shouldRun` method in the `AbstractRecordedMigration`.

Reply:
> As discussed in slack, in most cases you will need to implement your own shouldRun method anyway

Makes sense :+1: 

> Otherwise I am fine with providing a default implementation for the shouldRun method

Yeah, because you could *either* have an empty abstract method or a default implementation, I'd go with the default implementation. Then you get documentation how this feature works for free and adding your own stuff is just a quick `CTRL + o` (or similar) away. :slightly_smiling_face: